### ISSUE TITLE
User-defined functions in main modules without `local` prefix

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#657]
I wonder where this has been discussed before (but I can’t find it):

For simple scripts and main modules, the necessity to prefix all functions with a `local` prefix is cumbersome. Next, it’s counterintuitive as the prefix is not required for variable declarations:

```xquery
declare function local:f() { 1 };
declare variable $x := 1;
$x + local:f()
```

It’s possible to use `declare default function namespace '...';`, but that’s doesn’t feel much easier:

```xquery
declare default function namespace 'x';
declare function f() { 1 };
f()
```

I wonder if we can do one of the following things?

1. Allow functions without namespace (`declare function x() {}`, `declare function Q{}x() {}`).
2. Assign functions without namespace to the default function namespace.

**Comment by ChristianGruen** [comment]
> 2. Assign functions without namespace to the default function namespace.

…is probably not the best option; otherwise, function calls would need to be rewritten as well.

**Comment by ChristianGruen** [comment]
After a closer look, the least invasive option would probably be to allow users to define functions in the default `fn` namespace, as long as the function name does not conflict with the names of the system functions. Would that be possible?

```xquery
(: legal: name does not conflict :)
declare function inc($x) { $x + 1 };
inc(4)

(: illegal: name conflicts, arity does not matter :)
declare function count($x) { },
declare function count($x, $y) { };
```

I assume that most users won’t notice in which namespace the function is, or it won’t be relevant anyway.

**Comment by michaelhkay** [comment]
The problem (and the reason this wasn't allowed in the first place) is that when we later introduce a new function in the core library, we introduce a potential incompatibility for users.

I prefer an approach where the user-declared functions go in a different namespace (like local), but unprefixed function calls can search multiple namespaces (in a defined order). That way, adding a new fn:inc() function doesn't create a problem for a query that calls local:inc() with no prefix,  it only means that if the user wants to modify the query to call fn:inc(), they have to use a prefix to disambiguate it.

See issue #369 for a general discussion.

**Comment by benibela** [comment]
In Xidel I have  set a different default function namespace.  The spec says an implementation can use something else than `fn` as default. 

Then the user can just write:

    declare function count($x,$y){$x+$y}; 
    namespace-uri-from-QName(function-name(count#1)),
    namespace-uri-from-QName(function-name(count#2)),
    namespace-uri-from-QName(function-name(fn:count#1)),
    namespace-uri-from-QName(function-name(pxp:count#1))

    (: prints
    http://www.benibela.de/2012/pxp/extensions
    http://www.benibela.de/2012/pxp/extensions
    http://www.w3.org/2005/xpath-functions
    http://www.benibela.de/2012/pxp/extensions
    :)

**Comment by ChristianGruen** [comment]
@benibela Note that a different default function namespace also affects function calls. This expression must raise an error:

```xquery
declare default function namespace '_';
true()
```

**Comment by ChristianGruen** [comment]
> I prefer an approach where the user-declared functions go in a different namespace (like local), but unprefixed function calls can search multiple namespaces (in a defined order). That way, adding a new fn:inc() function doesn't create a problem for a query that calls local:inc() with no prefix, it only means that if the user wants to modify the query to call fn:inc(), they have to use a prefix to disambiguate it.

@michaelhkay +1 for `local`. I assume that an unprefixed user-declared function must only be assigned to this namespace if the default function namespace is `http://www.w3.org/2005/xpath-functions`. – Thanks for the pointer to #369.

**Comment by michaelhkay** [comment]
See also issue #757

**Comment by ChristianGruen** [comment]
> I prefer an approach where the user-declared functions go in a different namespace (like local), but unprefixed function calls can search multiple namespaces (in a defined order). 

After a second thought, I think that this approach would be problematic (unless I’m overlooking something):

In the following snippet, it’s obvious that the unprefixed function call invokes `fn:sum`:

```xquery
declare function local:sum($x) { };
sum(3)
```

In the next snippet, if the unprefixed `inc` function is assigned to the `local` namespace as suggested, and if the unprefixed `inc` function call is associated with `local:inc`, the code could be successfully parsed:

```xquery
declare function inc($n) { $n + 1 };
inc(3)
```

If an `fn:inc#1` function was added in a future version of the spec, wouldn’t the behavior of this code change?

Wouldn’t this be more drastic than rejecting the old code, as suggested in the initial comments of this thread?

**Comment by michaelhkay** [comment]
I wonder whether the following would work:

(a) Allow user-defined functions to be in no namespace (including private functions in library modules, which currently have to be in the module namespace)
(b) Extend the EQName syntax so that `:thing` represents a QName in no namespace with local name `thing` -- this being allowed wherever an EQName is permitted, including in function declarations and function calls.

```
declare function :increment($x) {$x + 1};
:increment(23)
```

We would have to look carefully for possible syntax ambiguities, for example in map constructors where `:` is used as a free-standing symbol.

**Comment by ChristianGruen** [comment]
Would it be an option to allow unprefixed function declarations and assign them to the `fn` namespace, but warn users (in the spec, and possibly via warnings output by an implementation) that such code may become incompatible with future versions of the language?

**Comment by michaelhkay** [comment]
It's an option, but I don't think it's a good idea.

**Comment by ChristianGruen** [comment]
Would it be possible to allow function declarations without namespace and to make the following function declarations equivalent?

```
declare function Q{}hope·dies·last() { };
declare function    hope·dies·last() { };
```

To enable `hope·dies·last()`, we would need to search function declarations without namespace and in the default function namespace.

**Comment by michaelhkay** [comment]
I think there's mileage in the idea of having an unprefixed function name result in a search of multiple namespaces. But the devil is in the detail.

**Comment by ChristianGruen** [comment]
With the arrival of record declarations, unprefixed function declarations would be even more interesting, in particular because otem type declarations need no prefix:

```xquery
declare type continent as enum('Africa', ...);
declare record why:coord(x, y);
```

**Comment by ChristianGruen** [comment]
> Technically, it is already possible:

Exactly, it is. With the latest changes, it is also possible for functions.

The issue can be closed (#2200 was accepted).
